extensions/amp-redbull-player/0.1/amp-redbull-player.js

```diff
@@ -0,0 +1,296 @@
+import {addParamsToUrl} from '../../../src/url';
+import {dict} from '../../../src/utils/object';
+import {disableScrollingOnIframe} from '../../../src/iframe-helper';
+import {getData, listen} from '../../../src/event-helper';
+import {getDataParamsFromAttributes, removeElement} from '../../../src/dom';
+import {installVideoManagerForDoc} from '../../../src/service/video-manager-impl';
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {userAssert} from '../../../src/log';
+
+/** @private @const */
+const TAG = 'amp-redbull-player';
+
+/** @private @const */
+const ANALYTICS_EVENT_TYPE_PREFIX = 'video-custom-';
+
+/** @private @const */
+const SANDBOX = [
+  SandboxOptions.ALLOW_SCRIPTS,
+  SandboxOptions.ALLOW_SAME_ORIGIN,
+  SandboxOptions.ALLOW_POPUPS,
+  SandboxOptions.ALLOW_POPUPS_TO_ESCAPE_SANDBOX,
+  SandboxOptions.ALLOW_TOP_NAVIGATION_BY_USER_ACTIVATION,
+];
+
+/** @implements {../../../src/video-interface.VideoInterface} */
+class AmpRedBullPlayer extends AMP.BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {?Element} */
+    this.iframe_ = null;
+
+    /** @private {!UnlistenDef|null} */
+    this.unlistenFrame_ = null;
+
+    /** @private {string} */
+    this.tagId_ = '';
+
+    /**
+     * @param {!Event} e
+     * @return {undefined}
+     * @private
+     */
+    this.boundOnMessage_ = e => this.onMessage_(e);
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  buildCallback() {
+    const {element} = this;
+
+    /*
+     * Required Paramaters: data-param-videoid
+     * Break creation if required parameters are missing!
+     */
+    userAssert(
+      element.getAttribute('data-param-videoid'),
+      'The data-param-videoid attribute is required for %s',
+      element
+    );
+
+    this.tagId_ = element.getAttribute('id') || 'rbvideo';
+
+    installVideoManagerForDoc(element);
+  }
+
+  /** @override */
+  layoutCallback() {
+    const {element} = this;
+
+    const params = getDataParamsFromAttributes(element);
+    const videoId = params['videoid'];
+    const skinId = params['skinid'] || 'com';
+    const locale = params['locale'] || 'global';
+
+    const origin = 'https://player.redbull.com/amp/amp-iframe.html';
+
+    const src = addParamsToUrl(
+      origin,
+      dict({
+        'videoId': videoId,
+        'skinId': skinId,
+        'ampTagId': this.tagId_,
+        'locale': locale,
+      })
+    );
+
+    this.iframe_ = disableScrollingOnIframe(
+      createFrameFor(this, src, '', SANDBOX)
+    );
+
+    this.unlistenFrame_ = listen(this.win, 'message', this.boundOnMessage_);
+    return this.loadPromise(this.iframe_).then(() => {
+      this.onReady_();
+    });
+  }
+
+  /** @private */
+  onReady_() {
+    Services.videoManagerForDoc(this.element).register(this);
+    const message = `{"msg":"amp-loaded","id":"${TAG}-${this.tagId_}"}`;
```

Comment:
`JSON.encode` here please.

Reply:
And JSON.encode here is. Thanks for reviewing!